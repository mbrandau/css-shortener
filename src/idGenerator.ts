// Leaving 'd' out to avoid generating the word 'ad'
const defaultAlphabet = 'abcefghijklmnopqrstuvwxyz0123456789_-';

export function createIdGenerator(alphabet?: string) {
  var options = {
    alphabet: alphabet || defaultAlphabet,
    length: 1,
    index: 0,
  };

  return function() {
    var res = generateId(options);
    // class names should not start with a number or a dash
    while (/^[0-9-].*$/.test(res)) res = generateId(options);
    return res;
  };
}

var generateId = function(options: {
  length: number;
  alphabet: string;
  index: number;
}) {
  var res = '';

  for (var i = options.length - 1; i >= 0; i--) {
    var x = Math.pow(options.alphabet.length, i);
    var n = Math.floor(options.index / x);
    res += options.alphabet[n % options.alphabet.length];
  }

  options.index++;
  if (options.index > Math.pow(options.alphabet.length, options.length) - 1) {
    options.length++;
    options.index = 0;
  }

  return res;
};
