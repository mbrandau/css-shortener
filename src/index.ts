import { createIdGenerator } from './idGenerator';

type Options = {
  alphabet?: string;
};

export type ClassNameMap = { [longName: string]: string };

export class CssShortener {
  private _map: ClassNameMap = {};
  private generateId: () => string;

  constructor(options?: Options) {
    this.generateId = createIdGenerator(options?.alphabet);
  }

  public get map(): ClassNameMap {
    return this._map;
  }

  /**
   * Imports mapped CSS class names
   * @param {object} map Map that should be imported
   * @param {boolean} override If true, existing mappings will be overridden
   */
  public importMap(map: ClassNameMap, override: boolean = false) {
    for (let classNameToImport in map) {
      if (this.isAlreadyUsed(map[classNameToImport]))
        throw new Error(
          `Cannot import class '${classNameToImport}'. Short class name '${map[classNameToImport]}' is already being used.`
        );
      if (this._map[classNameToImport] != null) {
        // Override mapped class name
        if (override === true)
          this._map[classNameToImport] = map[classNameToImport];
      } else this._map[classNameToImport] = map[classNameToImport]; // Import class name
    }
  }

  /**
   * Checks if the given class name is already used as a replacement for a longer class name
   * @param className class name to check for
   */
  private isAlreadyUsed(className: string): boolean {
    return Object.values(this.map).includes(className);
  }

  /**
   * Generates and maps a new class name
   * @param className The existing class name that should be replaced
   * @returns {string} The new generated and mapped class name
   */
  public shortenClassName(className: string) {
    // Return the already mapped class name
    if (this._map[className] != null) return this._map[className];
    // Generate, map and return the new class name
    let newClassName = this.generateId();
    while (this.isAlreadyUsed(newClassName)) {
      newClassName = this.generateId();
    }
    return (this._map[className] = newClassName);
  }
}
