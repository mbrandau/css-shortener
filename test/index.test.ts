import { CssShortener } from '../src';
import { createIdGenerator } from '../src/idGenerator';

describe('CssShortener', function() {
  it('should use custom alphabet', function() {
    const c = new CssShortener({
      alphabet: 'a',
    });
    expect(c.shortenClassName('class1')).toBe('a');
    expect(c.shortenClassName('class2')).toBe('aa');
    expect(c.shortenClassName('class3')).toBe('aaa');
    expect(c.map).toStrictEqual({
      class1: 'a',
      class2: 'aa',
      class3: 'aaa',
    });
  });
  describe('#importMap()', function() {
    it('should import mappings from a map object', function() {
      const c = new CssShortener(),
        map = {
          'test-class': 'a',
        };
      c.importMap(map);
      expect(c.map).toStrictEqual(map);
    });
    it('should not override existing mappings', function() {
      const c = new CssShortener(),
        prevMap = {
          'test-class': 'a',
        },
        importMap = {
          'test-class': 'b',
        };
      c.importMap(prevMap);
      c.importMap(importMap);
      expect(c.map).toStrictEqual(prevMap);
    });
    it('should override existing mappings', function() {
      const c = new CssShortener(),
        prevMap = {
          'test-class': 'a',
        },
        importMap = {
          'test-class': 'b',
        };
      c.importMap(prevMap);
      c.importMap(importMap, true);
      expect(c.map).toStrictEqual(importMap);
    });
    it('throws when a short name to import is already in the map', () => {
      const c = new CssShortener(),
        prevMap = {
          'test-class': 'a',
        },
        importMap = {
          'test-class2': 'a',
        };
      c.importMap(prevMap);
      expect(() => c.importMap(importMap)).toThrowErrorMatchingSnapshot();
    });
  });
  describe('#map', function() {
    it('should return a map with original class names and shortened versions', function() {
      const c = new CssShortener(),
        map = {
          'test-class': 'a',
        };
      c.importMap(map);
      expect(c.map).toStrictEqual(map);
    });
  });
  describe('#shortenClassName', function() {
    it('should not reuse already existing short class names', function() {
      const c = new CssShortener(),
        map = {
          'test-class': 'b',
        };
      c.importMap(map);
      expect(c.shortenClassName('test1')).toBe('a');
      expect(c.shortenClassName('test2')).toBe('c');
      expect(c.shortenClassName('test-class')).toBe('b');
    });
  });
});

describe('IdGenerator (default alphabet)', function() {
  const gen = createIdGenerator();
  it('should not start with a number or a hyphen', function() {
    for (let i = 0; i < 100; i++) {
      expect(gen()).not.toMatch(/^[0-9-].*$/);
    }
  });
});
