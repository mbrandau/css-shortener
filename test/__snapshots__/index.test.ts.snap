// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CssShortener #importMap() throws when a short name to import is already in the map 1`] = `"Cannot import class 'test-class2'. Short class name 'a' is already being used."`;
